### ISSUE TITLE
[Feature Request] Implement Head Related Impulse Respose (HRIR) support for the Convolver Effect

### ISSUE THREAD

**Issue by mastr-ch13f** [wwmm/easyeffects#2783]
Would it be possible to implement:
- up-mixing 2-ch audio into 5.1/7.1
-  process those channels through an 8-ch Impulse Response file then...
- down-mixing them back to stereo inside the Convolver alone?

From what I can understand, Zita-convolver library seems to support 64 audio channels in/out.

**Comment by violetmage** [comment]
What would be the purpose of using an HRIR/HRTF if you only have a stereo stream to upmix with?

**Comment by wwmm** [comment]
As far as I know the hrir file is supposed to be used when taking a multichannel input and processing it to generate a stereo audio that carries the surround experience. In EasyEffects we have only a stereo pipeline. I do not see how hrir files could be used.

**Comment by stefa168** [comment]
Pretty interested to see this feature implemented as well!

**Comment by TheBestKmanEver** [comment]
this can be achieved in openal. it takes .mhr files to run. you can use .sofa files or .def files for the creation of the HRTF with 'makemhr' (part of openal),

**Comment by wwmm** [comment]
> this can be achieved in openal. it takes .mhr files to run. you can use .sofa files or .def files for the creation of the HRTF with 'makemhr' (part of openal), shoot me a message if you want to know how to set this up.

But doesn't HRTF needs multichannel audio as input?

**Comment by TheBestKmanEver** [comment]
> > this can be achieved in openal. it takes .mhr files to run. you can use .sofa files or .def files for the creation of the HRTF with 'makemhr' (part of openal), shoot me a message if you want to know how to set this up.
> 
> But doesn't HRTF need multichannel audio as input?

nope. HRTF works based on how the human ear works with lots of math
HRIR is recorded from either dummy heads or human subjects, 1 mic in either ear with multiple sine sweeps recorded at various locations allowing you to have a full sphere of audio, HRTF uses Stereo for it's rendering as you're going to be using it on headphones anyway, everything gets mixed into HRTF, games may need you to enable a 3D audio feature to run, movies can work aswell, (has to be setup right in the player)

**Comment by wwmm** [comment]
>HRTF uses Stereo for it's rendering as you're going to be using it on headphones anyway, everything gets mixed into HRTF

Yes. But at least a few years ago when I tried to use HRTF in Pulseaudio the way to configure it was feeding  surround audio to the plugin so it created a stereo sound that simulated the surround experience that was coming from the input.

> HRTF is also able to make a mono signal sound like it is all around you

So what openal is doing in this case is simulating a surround sound without any relation to what is actually happening in the source material.  Not ideal but it may be useful in the right circumstances. Is there a code example somewhere?

**Comment by TheBestKmanEver** [comment]
https://github.com/kcat/openal-soft you can probably get more info here

**Comment by ewheelerinc** [comment]
+1

**Comment by alastair-dm** [comment]
https://lemmy.world/post/1865680 has instructions for a couple of ways to set it up in pipewire, both using pipewire filter-chains. The first uses its SOFA support and the second uses the convolver. It defines a 7.1 input and stereo output, but you could do a simpler version with just stereo in and out. The config for the SOFA version provides positions (angle, elevation, radius) for each of the sources - for stereo input you just need FL and FR which it puts at +/-30 degrees at head level. If you have a good HRTF for you then the effect is to make it sound like listening to a pair of speakers, instead of the in-head sound of normal stereo on headphones. Think of it as a much better version of the crossfeed plugin.

**Comment by wwmm** [comment]
>It defines a 7.1 input and stereo output, but you could do a simpler version with just stereo in and out. 

In our case it has to be the simpler version. More than 2 channels is not viable right now.

**Comment by alastair-dm** [comment]
> In our case it has to be the simpler version. More than 2 channels is not viable right now.

I know. That's why I explained why the method described is also applicable to stereo input with headphones as an improvement on crossfeed.

**Comment by LethalManBoob** [comment]
Hello. Recently i submitted a request to allow 4 channel convolver filters. They work just fine now on arch, i do not know if we have true stereo HRIR (LL/LR/RL/RR) and partitioned processing but we are quite far along this now.

**Comment by wwmm** [comment]
It isn't exactly what was requested here but we now have basic support for sofa impulse files in our convolver

<img width="1320" height="993" alt="Image" src="https://github.com/user-attachments/assets/96270a28-431a-4e7b-9e48-8a1a8d1f3e9a" />
. The dialog in the image can be used if the impulse is from a sofa file.

Being honest it was quite hard to extract the impulse response from the sofa file. So it may not be totally correct yet. This definitely has to be tested. People able to build packages from our master branch can do that.

**Comment by SergSel2006** [comment]
I hope this gets full support someday.
I guess there could be different approach: Modify easyeffects sink to be sofa spatializer sink in pipewire - it'll offload HTRF to it (pipewire knows how to do it, plus it may do it more efficient), and you'll get additional output where you connect other filters in chain. I kind of use it that way, and it works, it just needs a tick in settings honestly

**Comment by wwmm** [comment]
>I guess there could be different approach: Modify easyeffects sink to be sofa spatializer sink in pipewire - it'll offload HTRF to it (pipewire knows how to do it, plus it may do it more efficient), and you'll get additional output where you connect other filters in chain

I never used this sink. So I am not familiar with its workflow. Does it allow the hrtf application to be disabled? If yes does it allow that without the need to reload the sink? If possible I would like to avoid having different sinks for different tasks. And reloading them is also an annoyance. The ideal situation is to load the sink once and changing its parameters on the fly.

**Comment by SergSel2006** [comment]
> > I guess there could be different approach: Modify easyeffects sink to be sofa spatializer sink in pipewire - it'll offload HTRF to it (pipewire knows how to do it, plus it may do it more efficient), and you'll get additional output where you connect other filters in chain
> 
> I never used this sink. So I am not familiar with its workflow. Does it allow the hrtf application to be disabled? If yes does it allow that without the need to reload the sink? If possible I would like to avoid having different sinks for different tasks. And reloading them is also an annoyance. The ideal situation is to load the sink once and changing its parameters on the fly.

My initial idea was to make the easyeffects input sink that pipewire surround sink so that every application thinks it uses surround, but I now see what you need. I see no problems as of right now with it by the way, even stereo sounds will be playable fine because there's FL/FR as always. 
I don't know if it allows to change params, but there are some tutorials to make it work.
The one I used is this: https://lemmy.world/post/1865680 with sofa approach. There's also tutorials on ArchWiki for it on Pipewire page.
